### PR TITLE
_86Box: bring roms together

### PR DESCRIPTION
**Pull request by lucasew**
## Description of changes
- Bring the room repo together in a way that 86Box recognizes and works out of the box
- Fix problem about QT complaining a version conflict

## Test instructions
- Build it
- Open it, GUI should start right after
- By default it starts some kind of bios and drops you in a BASIC shell
- `print "it works"`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 288072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_86Box</li>
  </ul>
</details>

**Comment by jchv** [comment]
I appreciate what you have done here, however the 86box roms repository is unfree. I'm sure some PC BIOS ROMs were redistributable to some degree, but they are still considered unfree as they're not available under an open source license.

If you just want decent usability, I've found cloning the ROMs into the home directory, under `~/.local/share/86Box/`, is a good solution. Then the package works without needing to do anything strange.

I'm not sure what the best course of action is here. If it would be acceptable, maybe an input to the derivation could specify whether or not ROMs should be fetched and included, and also make sure the license is marked unfree, and then this could be set up to a new `_86Box-with-roms` package entry. That said, I would not feel confident that this is the right way to go, and it might need more discussion.

**Comment by lucasew** [comment]
Isn't it better to just put the package as unfree because AFAIK it's kinda useless without the roms.

**Comment by jchv** [comment]
> Isn't it better to just put the package as unfree? Because AFAIK it's kinda useless without the roms.

Not necessarily, there are open source BIOS implementations that I can imagine would work, but I feel there are more things at play than just that.

For one thing, I'm a little skeptical that the 86box `roms` repository is actually legally redistributing all of the copyrighted ROM files that they have. If they are redistributing ROM images that are not distributed under terms that allow for redistribution, and don't have explicit permission to be redistributing them, then as far as I know, legally speaking, they are illicit copies. This makes it a little questionable if there should really be a derivation that pulls them in.

For another, even if there were not open source ROM images available for any of the machines that would actually boot in 86box, I definitely wonder if it's still not more appropriate to offer 86box with ROMs as a separate derivation. A lot of people use nixpkgs in a lot of different ways, like for example, one can use `nix develop nixpkgs#_86Box` to get a development environment for building 86box. If you make the _main_ 86Box derivation unfree by forcing unfree blobs into it, that will cause it to throw an error by default and require a workaround.

There's a lot of ways this could be gone about. We could throw in a small shell script that just clones the correct version of the `roms` repo into the user's home directory as-needed. Obviously, this just moves the concern over, but AFAIK it does at least mean that the package doesn't need to be unfree. Seems not much less reasonable than e.g. what Winetricks does when it downloads packages.

But honestly, I have no idea what the right thing to do here is. I am not suggesting any amount of authority w.r.t. how NixOS and Nixpkgs handle situations like these. Frankly, I'd really like for someone with more authority to step in and give some clarity as to the precedents so far and what would be considered acceptable here.

File: pkgs/applications/emulators/86box/default.nix [hunk 1]
```diff
@@ -48,14 +49,26 @@ stdenv.mkDerivation rec {
...
+    rev = "v${version}";
+    hash = "sha256-1HtoizO0QIGNjQTW0clzRp40h1ulw55+iTYz12UJSms=";
+  });
+
+  postInstall = ''
+    if [ ! -z "$roms" ]; then
+      mkdir -p $out/share/86Box
+      ln -s $roms $out/share/86Box/roms
+    fi
+  '' + (lib.optionalString stdenv.isLinux ''
```
Comment (SuperSandro2000) [inline pkgs/applications/emulators/86box/default.nix:64]: The ( is superfluous here

**Review by SuperSandro2000 — APPROVED** [review]
Diff lgtm, didn't test it though

**Comment by SuperSandro2000** [comment]
> Isn't it better to just put the package as unfree? Because AFAIK it's kinda useless without the roms.

That doesn't matter for meta.license. I am sure there are other packages in nixpkgs which are marked as free but they are not very useful without some unfree addition. Still we should set the metadata correct.

**Comment by lucasew** [comment]
What's the veredict then?

CI is telling to put stuff in by-name but I think it's out of the scope for this PR right now.

**Review by jchv — APPROVED** [review]
Diff LGTM + tested and seems to work. Not 100% sure if this is a good idea but lacking any stronger opposition I don't really want to block it.

File: pkgs/applications/emulators/86box/default.nix [hunk 2]
```diff
@@ -67,6 +80,7 @@ stdenv.mkDerivation rec {
   in
   ''
     wrapProgram $out/bin/86Box \
+      --unset QT_PLUGIN_PATH \
       "''${qtWrapperArgs[@]}" \
       --prefix ${libPathVar} : "${libPath}"
   '';
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:86]: I also think it doesn't make sense this `--unset`. Also, although it's not code you wrote, I must say I would have implemented this additional `--prefix` path completely different - I'd use `makeWrapperArgs+=(...)` in a `preFixup`, and I wouldn't use `dontWrapQtApps`.

File: pkgs/applications/emulators/86box/default.nix [hunk 3]
```diff
@@ -48,14 +49,26 @@ stdenv.mkDerivation rec {
...
+    fi
+  '' + (lib.optionalString stdenv.isLinux ''
     install -Dm644 -t $out/share/applications $src/src/unix/assets/net.86box.86Box.desktop
 
     for size in 48 64 72 96 128 192 256 512; do
       install -Dm644 -t $out/share/icons/hicolor/"$size"x"$size"/apps \
         $src/src/unix/assets/"$size"x"$size"/net.86box.86Box.png
     done;
-  '';
+  '');
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:71]: Unnecessary here - see [this comment](https://github.com/NixOS/nixpkgs/pull/288072/files#r1486462166)

File: pkgs/applications/emulators/86box/default.nix [hunk 4]
```diff
@@ -48,14 +49,26 @@ stdenv.mkDerivation rec {
...
-  postInstall = lib.optional stdenv.isLinux ''
+  roms = lib.optionalString unfreeEnableRoms (fetchFromGitHub {
+    owner = "86Box";
+    repo = "roms";
+    rev = "v${version}";
+    hash = "sha256-1HtoizO0QIGNjQTW0clzRp40h1ulw55+iTYz12UJSms=";
+  });
+
+  postInstall = ''
+    if [ ! -z "$roms" ]; then
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:60]: This is terrible. Why not using Nix to perform such a check? How about using a `let in` predicate in the beginning of the expression, and set `roms` there unconditionally, and then in the `postInstall`, use `lib.optionalString unfreeEnableRoms`?

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -2613,6 +2613,10 @@ with pkgs;
 
   _86Box = callPackage ../applications/emulators/86box { };
 
+  _86Box-with-roms = callPackage ../applications/emulators/86box {
+    unfreeEnableRoms = true;
```
Comment (doronbehar) [inline pkgs/top-level/all-packages.nix:2617]: If you are making this package unfree and Hydra won't build it anyway, maybe you should also `unfreeEnableDiscord`? Also, I'd use `_86Box.override`.

File: pkgs/applications/emulators/86box/default.nix (hunk 3 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix:71]: I added to make it explicit

File: pkgs/applications/emulators/86box/default.nix (hunk 2 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix]: Done and working.

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (lucasew) [reply pkgs/top-level/all-packages.nix:2617]: Do you think it would be better to use `pkgs/by-name` instead then override?

File: pkgs/applications/emulators/86box/default.nix (hunk 3 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix:71]: Removed

File: pkgs/applications/emulators/86box/default.nix (hunk 4 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix]: It uses the version, I plan to make it compatible with r-ryantm.

File: pkgs/applications/emulators/86box/default.nix [hunk 6]
```diff
@@ -48,36 +49,42 @@ stdenv.mkDerivation rec {
...
 
-  postInstall = lib.optional stdenv.isLinux ''
+  postInstall = let
+    romsRepo = fetchFromGitHub {
+      owner = "86Box";
+      repo = "roms";
+      rev = "v${version}";
+      hash = "sha256-1HtoizO0QIGNjQTW0clzRp40h1ulw55+iTYz12UJSms=";
+    };
+  in (lib.optionalString stdenv.isLinux ''
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:59]: This is a bit better, but still a bit unreadable IMO, what I meant before is to use:

```nix
let
  romsRepo = ...
in stdenv.mkDerivation rec {
}
```

Another option, which is even better IMO, would be, to use:

```nix
stdenv.mkDerivation(finalAttrs: {
  ...
  passthru = {
    roms = fetchFromGitHub {
       ...
    };
  };

  postInstall = ''
  '' + lib.optionalString unfreeEnableRoms ''
    ln -s ${finalAttrs.passthru.roms} $out/...
  '';
```

This way, users who want to modify the roms derivation, could do it declarativly with:

```nix
my86box = _86box-with-roms.overrideAttrs:(finalAttrs: {
  passthru.roms = fetchFromGitHub {
    # Use my fork...
  };
};
```

I'm not totally sure about the above, so it is worth testing - you can insert that expression in `all-packages.nix` and try to build it and test it this way.

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (doronbehar) [reply pkgs/top-level/all-packages.nix:2617]: > Do you think it would be better to use `pkgs/by-name` instead then override?

That's unrelated to one another. You won't be able to use `pkgs/by-name` trivially for an expression that is equal to `_86box-with-roms.override { ... };`.

I'd let the automatic transition to `pkgs/by-name` do the job.

**Review by doronbehar — CHANGES_REQUESTED** [review]
Changes look better, but I'd split the changes to 2 different commits:

```
_86box: use preFixup and `makeWrapperArgs` instead of `postFixup`
_86box: support unfree roms
```

If you'd use the `stdenv.mkDerivation(finalAttrs: {})` style, that should go to a separate commit as well.

File: pkgs/applications/emulators/86box/default.nix (hunk 6 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix:59]: nice

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 288072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_86Box</li>
    <li>_86Box-with-roms</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Derivation didn't change. No build was made during this `nixpkgs-review` run.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 288072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_86Box</li>
    <li>_86Box-with-roms</li>
  </ul>
</details>

File: pkgs/applications/emulators/86box/default.nix (hunk 6 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix]: I did the last one

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 288072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_86Box</li>
    <li>_86Box-with-roms</li>
  </ul>
</details>

File: pkgs/applications/emulators/86box/default.nix [hunk 7]
```diff
@@ -48,38 +49,44 @@ stdenv.mkDerivation rec {
...
-  postInstall = lib.optional stdenv.isLinux ''
+  passthru = {
+    roms = fetchFromGitHub {
+      owner = "86Box";
+      repo = "roms";
+      rev = "v${finalAttrs.version}";
+      hash = "sha256-1HtoizO0QIGNjQTW0clzRp40h1ulw55+iTYz12UJSms=";
+    };
+  };
+  postInstall = (lib.optionalString stdenv.isLinux ''
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:60]: Unnecessary `(` added, already pointed out by @SuperSandro2000 in [this comment](https://github.com/NixOS/nixpkgs/pull/288072/files#r1486462166) and in [my comment](https://github.com/NixOS/nixpkgs/pull/288072/files#r1533394386):

```suggestion
  postInstall = lib.optionalString stdenv.isLinux ''
```

File: pkgs/applications/emulators/86box/default.nix [hunk 8]
```diff
@@ -6,16 +6,17 @@
 , enableNewDynarec ? enableDynarec && stdenv.hostPlatform.isAarch
 , enableVncRenderer ? false
 , unfreeEnableDiscord ? false
+, unfreeEnableRoms ? false
 }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:12]: This change should be in it's own commit, before you add support for the unfree roms.

File: pkgs/applications/emulators/86box/default.nix [hunk 9]
```diff
@@ -48,38 +49,44 @@ stdenv.mkDerivation rec {
...
+  };
+  postInstall = (lib.optionalString stdenv.isLinux ''
     install -Dm644 -t $out/share/applications $src/src/unix/assets/net.86box.86Box.desktop
 
     for size in 48 64 72 96 128 192 256 512; do
       install -Dm644 -t $out/share/icons/hicolor/"$size"x"$size"/apps \
         $src/src/unix/assets/"$size"x"$size"/net.86box.86Box.png
     done;
-  '';
+  '') + (lib.optionalString unfreeEnableRoms ''
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:67]: For example here, without the extra `)`, this would like like this:

```suggestion
  '' + lib.optionalString unfreeEnableRoms ''
```

File: pkgs/applications/emulators/86box/default.nix [hunk 10]
```diff
@@ -48,38 +49,44 @@ stdenv.mkDerivation rec {
...
+    mkdir -p $out/share/86Box
+    ln -s ${finalAttrs.passthru.roms} $out/share/86Box/roms
+  '');
 
   # Some libraries are loaded dynamically, but QLibrary doesn't seem to search
   # the runpath, so use a wrapper instead.
-  postFixup = let
+  preFixup = let
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:74]: Another change that should be in a separate commit, as I asked before.

File: pkgs/top-level/all-packages.nix [hunk 11]
```diff
@@ -2620,6 +2620,11 @@ with pkgs;
 
   _86Box = callPackage ../applications/emulators/86box { };
 
+  _86Box-with-roms = callPackage ../applications/emulators/86box {
```
Comment (doronbehar) [inline pkgs/top-level/all-packages.nix:2623]: Use `.override` please, I asked that already.

**Review by doronbehar — CHANGES_REQUESTED** [review]
Many of the comments from the precious review weren't addressed. I repeated myself here.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 288072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_86Box</li>
    <li>_86Box-with-roms</li>
  </ul>
</details>

File: pkgs/applications/emulators/86box/default.nix (hunk 8 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix:12]: Done

File: pkgs/applications/emulators/86box/default.nix (hunk 7 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix]: Done

File: pkgs/applications/emulators/86box/default.nix (hunk 10 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix:77]: I did it with the first commit

File: pkgs/top-level/all-packages.nix (hunk 11 above)
Comment (lucasew) [reply pkgs/top-level/all-packages.nix:2623]: OK

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 288072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_86Box</li>
    <li>_86Box-with-roms</li>
  </ul>
</details>

File: pkgs/applications/emulators/86box/default.nix [hunk 12]
```diff
@@ -48,38 +49,47 @@ stdenv.mkDerivation rec {
     ++ lib.optional (!enableDynarec) "-DDYNAREC=OFF"
     ++ lib.optional (!unfreeEnableDiscord) "-DDISCORD=OFF";
 
-  postInstall = lib.optional stdenv.isLinux ''
-    install -Dm644 -t $out/share/applications $src/src/unix/assets/net.86box.86Box.desktop
+  postInstall =
+    lib.optionalString stdenv.isLinux ''
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:53]: Unnecessary indentation, that enlarges the diff.

File: pkgs/applications/emulators/86box/default.nix [hunk 13]
```diff
@@ -48,38 +49,47 @@ stdenv.mkDerivation rec {
...
-  '';
+      for size in 48 64 72 96 128 192 256 512; do
+        install -Dm644 -t $out/share/icons/hicolor/"$size"x"$size"/apps \
+          $src/src/unix/assets/"$size"x"$size"/net.86box.86Box.png
+      done;
+    ''
+  + lib.optionalString unfreeEnableRoms ''
+      mkdir -p $out/share/86Box
+      ln -s ${finalAttrs.passthru.roms} $out/share/86Box/roms
+    '';
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:64]: Here too.

File: pkgs/applications/emulators/86box/default.nix (hunk 10 above)
Comment (doronbehar) [reply pkgs/applications/emulators/86box/default.nix:77]: > I did it with the first commit

Please separate the changes into different commits.

**Review by doronbehar — CHANGES_REQUESTED** [review]
Please separate the changes into 3 commits:

1. `finalAttrs`
2. Better wrapping via: `postFixup` -> `preFixup`.
3. support `unfreeEnableRoms`.

File: pkgs/applications/emulators/86box/default.nix (hunk 12 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix]: Solved

File: pkgs/applications/emulators/86box/default.nix (hunk 13 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix]: Solved

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 288072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_86Box</li>
    <li>_86Box-with-roms</li>
  </ul>
</details>

File: pkgs/applications/emulators/86box/default.nix [hunk 14]
```diff
@@ -48,38 +49,46 @@ stdenv.mkDerivation rec {
...
     mainProgram = "86Box";
     homepage = "https://86box.net/";
-    license = with licenses; [ gpl2Only ] ++ optional unfreeEnableDiscord unfree;
+    license = with licenses; [ gpl2Only ]
+      ++ optional (unfreeEnableDiscord || unfreeEnableRoms) unfree;
     maintainers = [ maintainers.jchw ];
     platforms = platforms.linux;
   };
-}
+})
```
Comment (doronbehar) [inline pkgs/applications/emulators/86box/default.nix:94]: Sorry for being so strict, but this should be part of the `finalAttrs` commit. Also, the commit message of that is a bit long. Can become more concise.

**Review by doronbehar — COMMENTED** [review]
Seems almost perfect!

File: pkgs/applications/emulators/86box/default.nix (hunk 14 above)
Comment (lucasew) [reply pkgs/applications/emulators/86box/default.nix:94]: Done

**Review by doronbehar — APPROVED** [review]
